# Aprendizagem Automática Avançada - Assignment 5
# Cláudia Afonso nº 36273 & Rita Rodrigues nº 54859

## Problem 1

**Given the bayesian network provided in Figure 1, calculate the following queries:**

In [18]:
from pomegranate import *

In [19]:
# Probability of A
a_table = DiscreteDistribution({'T': 0.3, 'F': 0.7})

In [20]:
# Probability of B given A
b_table = ConditionalProbabilityTable(
        [['F', 'T', 0.40],
         ['F', 'F', 0.60],
         ['T', 'T', 0.80], 
         ['T', 'F', 0.20]], [a_table])

In [21]:
# Probability of D given B
d_table = ConditionalProbabilityTable(
        [['F', 'T', 0.55],
         ['F', 'F', 0.45],
         ['T', 'T', 0.20], 
         ['T', 'F', 0.80]], [b_table])

In [22]:
# Probability of C given A and B
c_table = ConditionalProbabilityTable(
        [['F','F','T', 0.10],
         ['F','F','F', 0.90],
         ['F', 'T', 'T', 0.70],
         ['F', 'T', 'F', 0.30],
         ['T', 'F', 'T', 0.50],
         ['T', 'F', 'F', 0.50],
         ['T', 'T', 'T', 0.99],
         ['T', 'T', 'F', 0.01]], [a_table, b_table])

In [23]:
# Construct the network
a_node = Node(a_table, name = 'A')
b_node = Node(b_table, name = 'B')
c_node = Node(c_table, name = 'C')
d_node = Node(d_table, name = 'D')

In [24]:
model = BayesianNetwork()
model.add_states(a_node, b_node, c_node, d_node)
model.add_edge(a_node, b_node)
model.add_edge(a_node, c_node)
model.add_edge(b_node, c_node)
model.add_edge(b_node, d_node)
model.bake()

**a) P (A = T | C = T, D = T)**

In [25]:
# P(A=T|C=T, D=T) = 0.505
model.predict_proba({'C':'T', 'D':'T'})
# alternative: model.predict_proba([[None, None, 'T', 'T']])

array([{
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "T" : 0.5054138717420109,
                   "F" : 0.49458612825798914
               }
           ],
           "frozen" : false
       }                                    ,
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.3516850950686168,
                   "T" : 0.6483149049313832
               }
           ],
           "frozen" : false
       }                                    , 'T', 'T'], dtype=object)

The probability of A being true, given that C and D are true, is approximately 50.5%.

**b) P (A = T | D = F)**

In [29]:
# P(A=T|D=F)
model.predict_proba({'D':'F'})
# alternative: model.predict_proba([[None, None, None, 'F']])

array([{
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "T" : 0.34651898734177244,
                   "F" : 0.6534810126582277
               }
           ],
           "frozen" : false
       }                                     ,
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.34177215189873406,
                   "T" : 0.6582278481012659
               }
           ],
           "frozen" : false
       }                                     ,
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.3915454254125941,
                   "T" : 0.6084545745874058
               }
           ],

The probability of A being true, given that D is false, is approximately 34.7%

**c) P (B = T | C = T)**

In [30]:
# P(B=T|C=T)
model.predict_proba({'C':'T'})
# alternative: model.predict_proba([[None, None, 'T', None]])

array([{
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "T" : 0.5253830090147861,
                   "F" : 0.4746169909852139
               }
           ],
           "frozen" : false
       }                                    ,
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.1899156736574448,
                   "T" : 0.8100843263425553
               }
           ],
           "frozen" : false
       }                                    , 'T',
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.7335295142199177,
                   "T" : 0.2664704857800823
               }
           ]

The probability of B being true, given that C is true, is approximately 81%

**d) P (B = T | A = T, C = T)**

In [32]:
# P(B=T|A=T, C=T)
model.predict_proba({'A':'T', 'C':'T'})
# alternative: model.predict_proba([[T, None, 'T', None]])

array(['T', {
                "class" : "Distribution",
                "dtype" : "str",
                "name" : "DiscreteDistribution",
                "parameters" : [
                    {
                        "F" : 0.11210762331838604,
                        "T" : 0.8878923766816139
                    }
                ],
                "frozen" : false
            }                                     , 'T',
       {
           "class" : "Distribution",
           "dtype" : "str",
           "name" : "DiscreteDistribution",
           "parameters" : [
               {
                   "F" : 0.7607623318385648,
                   "T" : 0.23923766816143516
               }
           ],
           "frozen" : false
       }                                    ], dtype=object)

The probability of B being true, given that A and C are true, is approximately 88.8%.

**e) P (C = T | A = F, B = F, D = F)**

In [33]:
# P(C=T|A=F, B=F, D=F)
model.predict_proba({'A':'F', 'B':'F', 'D':'F'})
# alternative: model.predict_proba([['F', 'F', None, 'F']])

array(['F', 'F', {
                     "class" : "Distribution",
                     "dtype" : "str",
                     "name" : "DiscreteDistribution",
                     "parameters" : [
                         {
                             "F" : 0.8999999999999999,
                             "T" : 0.10000000000000016
                         }
                     ],
                     "frozen" : false
                 }                                    , 'F'], dtype=object)

The probability of C being true, given that A, B and D are false, is approximately 10.0%.

## Problem 2

The following code uses the scikit-learn library to explore the performance of two different classifiers, Naive Bayes and Random Forest, on the breast cancer dataset.

For Naive Bayes, we explore the effect of different prior probabilities on the model's performance. Specifically, we define three different sets of prior probabilities: [0.3, 0.7], [0.5, 0.5], and [0.7, 0.3]. These prior probabilities represent different levels of belief or uncertainty about the two classes.

For Random Forest, we explore the effect of different combinations of the n_estimators and max_depth parameters. The n_estimators parameter controls the number of decision trees that are generated by the Random Forest algorithm. We explore three values for this parameter: 50, 100, and 150. The max_depth parameter controls the maximum depth of each decision tree in the Random Forest. We explore three values for this parameter: 3, 5, and 7. 

By exploring these different parameter combinations, we can gain insight into how the Naive Bayes and Random Forest classifiers perform on the breast cancer dataset, and which parameter values lead to the most accurate predictions.

In [7]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load the breast cancer dataset
data = load_breast_cancer()
X = data.data
y = data.target

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the parameter values to explore
nb_priors = [[0.3, 0.7], [0.5, 0.5], [0.7, 0.3]]
rf_estimators = [50, 100, 150]
rf_max_depth = [3, 5, 7]

# Create an empty DataFrame to store the results
results_df = pd.DataFrame(columns=["Classifier", "Parameters", "Accuracy"])

# Naive Bayes
for priors in nb_priors:
    gnb = GaussianNB(priors=priors)
    gnb.fit(X_train, y_train)
    y_pred_gnb = gnb.predict(X_test)
    acc_gnb = accuracy_score(y_test, y_pred_gnb)
    results_df = pd.concat([results_df, pd.DataFrame({"Classifier": ["Naive Bayes"],
                                    "Parameters": [f"Priors={priors}"],
                                    "Accuracy": [acc_gnb]})])

# Random Forest
for n_estimators in rf_estimators:
    for max_depth in rf_max_depth:
        rf = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth)
        rf.fit(X_train, y_train)
        y_pred_rf = rf.predict(X_test)
        acc_rf = accuracy_score(y_test, y_pred_rf)
        results_df = pd.concat([results_df, pd.DataFrame({"Classifier": ["Random Forest"],
                                        "Parameters": [f"n_estimators={n_estimators}, max_depth={max_depth}"],
                                        "Accuracy": [acc_rf]})])

# Print the results
results_df

Classifier                     Parameters  Accuracy
0    Naive Bayes              Priors=[0.3, 0.7]  0.964912
0    Naive Bayes              Priors=[0.5, 0.5]  0.973684
0    Naive Bayes              Priors=[0.7, 0.3]  0.964912
0  Random Forest   n_estimators=50, max_depth=3  0.964912
0  Random Forest   n_estimators=50, max_depth=5  0.964912
0  Random Forest   n_estimators=50, max_depth=7  0.964912
0  Random Forest  n_estimators=100, max_depth=3  0.964912
0  Random Forest  n_estimators=100, max_depth=5  0.964912
0  Random Forest  n_estimators=100, max_depth=7  0.964912
0  Random Forest  n_estimators=150, max_depth=3  0.964912
0  Random Forest  n_estimators=150, max_depth=5  0.964912
0  Random Forest  n_estimators=150, max_depth=7  0.964912

For Naive Bayes, we see that the model performs well on all three sets of prior probabilities, with accuracies ranging from 0.965 to 0.974. This suggests that the Naive Bayes classifier is not particularly sensitive to the choice of prior probabilities for this dataset.

For Random Forest, we see that the model's accuracy remains constant across all parameter combinations, 0.965. This suggests that the performance of the Random Forest classifier is not strongly affected by the number of decision trees or the maximum depth of the trees for this dataset.

Overall, these results suggest that both Naive Bayes and Random Forest classifiers are well-suited for the breast cancer dataset.